server/raft.go

```diff
@@ -2563,22 +2597,63 @@ func (n *raft) readTermVote() (term uint64, voted string, err error) {
 	return term, voted, nil
 }
 
-// writeTermVote will record the largest term and who we voted for to stable storage.
-// Lock should be held.
-func (n *raft) writeTermVote() error {
+func (n *raft) setWriteErr(err error) {
+	n.Lock()
+	defer n.Unlock()
+	n.werr = err
+}
+
+func (n *raft) fileWriter() {
+	s := n.s
+	defer s.grWG.Done()
+
+	n.RLock()
 	tvf := path.Join(n.sd, termVoteFile)
-	if _, err := os.Stat(tvf); err != nil && !os.IsNotExist(err) {
-		return err
+	psf := path.Join(n.sd, peerStateFile)
+	n.RUnlock()
+
+	for s.isRunning() {
+		select {
+		case <-n.quit:
+			return
+		case <-n.wtvch:
+			var buf [termVoteLen]byte
+			n.RLock()
+			copy(buf[0:], n.wtv)
+			n.RUnlock()
+			if err := ioutil.WriteFile(tvf, buf[:], 0640); err != nil {
+				n.setWriteErr(err)
+			}
+		case <-n.wpsch:
+			n.RLock()
+			buf := append(n.wps[:0:0], n.wps...)
+			n.RUnlock()
+			if err := ioutil.WriteFile(psf, buf, 0640); err != nil {
+				n.setWriteErr(err)
+			}
+		}
 	}
+}
+
+// writeTermVote will record the largest term and who we voted for to stable storage.
+// Lock should be held.
+func (n *raft) writeTermVote() {
 	var buf [termVoteLen]byte
 	var le = binary.LittleEndian
 	le.PutUint64(buf[0:], n.term)
-	// FIXME(dlc) - NoVote
 	copy(buf[8:], n.vote)
-	if err := ioutil.WriteFile(tvf, buf[:8+len(n.vote)], 0644); err != nil {
-		return err
+	b := buf[:8+len(n.vote)]
+
+	// If same as what we have we can ignore.
+	if bytes.Equal(n.wtv, b) {
+		return
+	}
+	// Stamp latest and kick writer.
+	n.wtv = b
+	select {
+	case n.wtvch <- struct{}{}:
```

Comment:
With being unbuffered, are you ok losing it if the fileWriter routine is not actively receiving?

Reply:
Ah I see what you mean, maybe should make at least 1 buffered.